source/device_defender.c

```diff
@@ -269,6 +301,90 @@ static int s_get_metric_report_json(
+                case DD_METRIC_NUMBER:
+                    cJSON_AddNumberToObject(item, "number", (double)custom_metrics_data[metric_index].data.number);
+                    break;
+                case DD_METRIC_NUMBER_LIST:
+                    list_size = aws_array_list_length(&custom_metrics_data[metric_index].data.list);
+                    json_list = cJSON_CreateArray();
+                    if (NULL == json_list) {
+                        goto cleanup;
+                    }
+                    cJSON_AddItemToObject(item, "number_list", json_list);
+                    for (size_t num_index = 0; num_index < list_size; ++num_index) {
+                        int64_t number = 0;
+                        aws_array_list_get_at(&custom_metrics_data[metric_index].data.list, &number, num_index);
+                        array_item = cJSON_CreateNumber((double)number);
+                        cJSON_AddItemToArray(json_list, array_item);
+                    }
+                    break;
+                case DD_METRIC_STRING_LIST:
+                    list_size = aws_array_list_length(&custom_metrics_data[metric_index].data.list);
+                    json_list = cJSON_CreateArray();
+                    if (NULL == json_list) {
+                        goto cleanup;
+                    }
+                    cJSON_AddItemToObject(item, "string_list", json_list);
+                    for (size_t string_index = 0; string_index < list_size; ++string_index) {
+                        struct aws_string *string_value = NULL;
+                        aws_array_list_get_at(
+                            &custom_metrics_data[metric_index].data.list, &string_value, string_index);
+                        array_item = cJSON_CreateString(aws_string_c_str(string_value));
+                        cJSON_AddItemToArray(json_list, array_item);
+                    }
+                    break;
+                case DD_METRIC_IP_LIST:
+                    list_size = aws_array_list_length(&custom_metrics_data[metric_index].data.list);
+                    json_list = cJSON_CreateArray();
+                    if (NULL == json_list) {
+                        goto cleanup;
+                    }
+                    cJSON_AddItemToObject(item, "ip_list", json_list);
+                    for (size_t ip_index = 0; ip_index < list_size; ++ip_index) {
+                        struct aws_string *ip_value = NULL;
+                        aws_array_list_get_at(&custom_metrics_data[metric_index].data.list, &ip_value, ip_index);
+                        array_item = cJSON_CreateString(aws_string_c_str(ip_value));
+                        cJSON_AddItemToArray(json_list, array_item);
+                    }
+                    break;
+                case DD_METRIC_UNKNOWN:
+                default:
+                    /* output warning or error */
+                    continue;
+                    break;
```

Comment:
If using all of the functions correctly, it should be unreachable. If someone managed to trample over some memory and modified the metrics->type value to something outside the enum range, set it to 0, or created a value and forgot to populate an extra case (without the compiler error) this would cover it.

Reply:
I added handing here to just log it.